source/android-source/pilgrimunitysdk/src/main/java/com/foursquare/pilgrimunitysdk/Utils.java

```diff
@@ -28,43 +28,71 @@
     private static final String KEY_STARTED = "started";
 
     private Utils() {
+
     }
 
-    static PilgrimUserInfo fromMap(Map<String, ?> map) {
-        if (map.isEmpty()) {
+    static String jsonFromUserInfo(@NonNull PilgrimUserInfo userInfo) {
+        JSONArray keys = new JSONArray();
+        JSONArray values = new JSONArray();
+
+        for (Map.Entry<String, String> entry : userInfo.entrySet()) {
+            keys.put(entry.getKey());
+            values.put(entry.getValue());
+        }
+
+        try {
+            JSONObject json = new JSONObject();
+            json.put("keys", keys);
+            json.put("values", values);
+            return json.toString();
+        } catch (JSONException e) {
             return null;
         }
+    }
+
+    static PilgrimUserInfo userInfoFromJSON(@NonNull String jsonString) {
```

Comment:
on the flipside of the above comment, could read directly into the jsonobject and youd already have the keys and the values

Reply:
I get the above comment but am confused about this one, doesn't the line`JSONObject json = new JSONObject(jsonString);` do what you're saying?

Reply:
yeah, comment was just that you would no longer have "keys" and "values" array. It would just be to loop through the keys in the `json` object directly.

Reply:
Actually that won't work, the above comment too, because what I do is take the user info dictionary on the C# side and turn it into 2 arrays with the keys and values stored in the separate arrays and send that to java using the `ISerializationCallbackReceiver` protocol.  It's kinda ugly but this is to overcome the unity limitation of not being able to serialize/deserialize unstructured data/arbitrary dictionaries.  https://github.com/foursquare/pilgrim-unity-sdk/pull/4/files#diff-67da2bbb70bb08e50d93fa7ff1d04d2aR89

Reply:
This is the way they recommend to do this in their docs: https://docs.unity3d.com/ScriptReference/ISerializationCallbackReceiver.html, so the json would for example look like
```
{
"keys":["gender","userId"],
"values":["male","12345"]
}
```

Reply:
ah, so you just have to assume that the indicies of the arrays match up.

Reply:
Yeah i add a check in the java/objc and return null if they don't, shouldn't happen and not much else can be done i don't think
https://github.com/foursquare/pilgrim-unity-sdk/pull/4/files#diff-e3d1d648ad16456d1b6431946308325dR59
https://github.com/foursquare/pilgrim-unity-sdk/pull/4/files#diff-618c91203b78e854975d6c4322f465e9R307